File: ga/gpt3_5_result_2024_6/result/1968F_cpp_gpt/2_1968F.cpp
use std::io::{self, Read};

fn solve() {
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    let mut iter = input.trim().split_whitespace();

    let n: usize = iter.next().unwrap().parse().unwrap();
    let q: usize = iter.next().unwrap().parse().unwrap();

    input.clear();
    io::stdin().read_line(&mut input).unwrap();
    let arr: Vec<i32> = input.split_whitespace().map(|x| x.parse().unwrap()).collect();

    let mut x: Vec<(i32, usize)> = vec![(0, 0); n + 1];
    x[0] = (0, 0);

    for i in 1..=n {
        x[i] = (x[i - 1].0 ^ arr[i - 1], i);
    }

    let mut x2: Vec<i32> = x.iter().map(|(val, _)| *val).collect();
    x.sort();

    for _ in 0..q {
        input.clear();
        io::stdin().read_line(&mut input).unwrap();
        let mut iter = input.trim().split_whitespace();

        let l: usize = iter.next().unwrap().parse().unwrap();
        let r: usize = iter.next().unwrap().parse().unwrap();

        let c = x2[r] ^ x2[l - 1];
        if c == 0 {
            println!("YES");
            continue;
        }

        let i1 = x.iter().position(|&(val, idx)| idx >= l && val == c ^ x2[l - 1]);

        match i1 {
            None => {
                println!("NO");
                continue;
            }
            Some(i) => {
                if x[i].1 > r {
                    println!("NO");
                    continue;
                }
            }
        }

        let i2 = x.iter().position(|&(val, idx)| idx >= x[i1.unwrap()].1 + 1 && val == x2[l - 1]);

        match i2 {
            None => {
                println!("NO");
            }
            Some(i) => {
                if x[i].1 > r {
                    println!("NO");
                } else {
                    println!("YES");
                }
            }
        }
    }
}

fn main() {
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    let t: usize = input.trim().parse().unwrap();

    for _ in 0..t {
        solve();
    }
}

File: ga/glm4_result_2024_6/result/1968F_cpp_glm/2_1968F.rs
use std::io::{self, BufRead};

fn solve() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines();

    let nq: Vec<usize> = lines.next().unwrap().unwrap().split_whitespace().map(|s| s.parse().unwrap()).collect();
    let n = nq[0];
    let q = nq[1];

    let arr: Vec<usize> = lines.next().unwrap().unwrap().split_whitespace().map(|s| s.parse().unwrap()).collect();

    let mut x = Vec::with_capacity(n + 1);
    x.push((0, 0));
    for (i, &a) in arr.iter().enumerate() {
        x.push((x[i].0 ^ a, i + 1));
    }
    let mut x2 = Vec::with_capacity(n + 1);
    for &(_, idx) in &x {
        x2.push(x[idx].0);
    }
    x2.sort();

    for _ in 0..q {
        let lr: Vec<usize> = lines.next().unwrap().unwrap().split_whitespace().map(|s| s.parse().unwrap()).collect();
        let l = lr[0];
        let r = lr[1];

        let c = x2[r] ^ x2[l - 1];
        if c == 0 {
            println!("YES");
            continue;
        }

        let i1 = match x2.binary_search(&c ^ x2[l - 1]) {
            Ok(i) => i,
            Err(i) => i,
        };

        if i1 == n || x[i1].1 > r || x[i1].0 != (c ^ x2[l - 1]) {
            println!("NO");
            continue;
        }

        let i2 = match x2.binary_search(&x2[l - 1]) {
            Ok(i) => i,
            Err(i) => i,
        };

        if i2 == n || x[i2].1 > r || x[i2].0 != x2[l - 1] {
            println!("NO");
        } else {
            println!("YES");
        }
    }
}

fn main() {
    let t = 1; // Read from input if needed
    for _ in 0..t {
        solve();
    }
}